### PR TITLE
A0-1811: `snark_relation` macro

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
This PR introduces `snark_relation` procedural macro for concise defining SNARK relations. For detailed documentation, consult code docs. For an example, see adaptation for xor relation or 

Given such definition:
```rust
use ark_std::{One, Zero};
use snark_relation_proc_macro::snark_relation;

use crate::CircuitField;

fn parse_u16(x: u16) -> CircuitField {
    CircuitField::from(x)
}

fn byte_to_bits<F: Zero + One + Copy>(byte: &u8) -> Vec<F> {
    let mut bits = [F::zero(); 8];
    for (idx, bit) in bits.iter_mut().enumerate() {
        if (byte >> idx) & 1 == 1 {
            *bit = F::one();
        }
    }
    bits.to_vec()
}

#[snark_relation]
mod relation {
    #[circuit_field]
    pub type CF = CircuitField;

    #[relation_object_definition]
    struct SomeRelation {
        #[constant]
        a: u8,
        #[public_input(order = "1", frontend_type = "u16", parse_with = "parse_u16")]
        b: CF,
        #[private_input(frontend_type = "u32")]
        c: u64,
        #[public_input(order = "0", serialize_with = "byte_to_bits")]
        d: u8,
    }

    #[circuit_definition]
    fn generate_circuit() -> ark_relations::r1cs::Result<()> {
        Ok(())
    }
}
```

the macro will expand to:
```rust
use ark_std::{One, Zero};
use snark_relation_proc_macro::snark_relation;
use crate::CircuitField;
fn parse_u16(x: u16) -> CircuitField {
    CircuitField::from(x)
}
fn byte_to_bits<F: Zero + One + Copy>(byte: &u8) -> Vec<F> {
    let mut bits = [F::zero(); 8];
    for (idx, bit) in bits.iter_mut().enumerate() {
        if (byte >> idx) & 1 == 1 {
            *bit = F::one();
        }
    }
    bits.to_vec()
}
pub type CF = CircuitField;
pub struct SomeRelationWithoutInput {
    a: u8,
}
impl SomeRelationWithoutInput {
    pub fn new(a: u8) -> Self {
        Self { a }
    }
    pub fn a(&self) -> &u8 {
        &self.a
    }
    pub fn b(
        &self,
    ) -> core::result::Result<&CF, ark_relations::r1cs::SynthesisError> {
        Err(ark_relations::r1cs::SynthesisError::AssignmentMissing)
    }
    pub fn d(
        &self,
    ) -> core::result::Result<&u8, ark_relations::r1cs::SynthesisError> {
        Err(ark_relations::r1cs::SynthesisError::AssignmentMissing)
    }
    pub fn c(
        &self,
    ) -> core::result::Result<&u64, ark_relations::r1cs::SynthesisError> {
        Err(ark_relations::r1cs::SynthesisError::AssignmentMissing)
    }
}
pub struct SomeRelationWithPublicInput {
    a: u8,
    b: CF,
    d: u8,
}
impl SomeRelationWithPublicInput {
    pub fn new(a: u8, b: u16, d: u8) -> Self {
        Self { a, b: parse_u16(b), d }
    }
    pub fn a(&self) -> &u8 {
        &self.a
    }
    pub fn b(
        &self,
    ) -> core::result::Result<&CF, ark_relations::r1cs::SynthesisError> {
        Ok(&self.b)
    }
    pub fn d(
        &self,
    ) -> core::result::Result<&u8, ark_relations::r1cs::SynthesisError> {
        Ok(&self.d)
    }
    pub fn c(
        &self,
    ) -> core::result::Result<&u64, ark_relations::r1cs::SynthesisError> {
        Err(ark_relations::r1cs::SynthesisError::AssignmentMissing)
    }
    pub fn serialize_public_input(&self) -> ark_std::vec::Vec<CF> {
        [
            byte_to_bits(&self.d),
            <[_]>::into_vec(#[rustc_box] ::alloc::boxed::Box::new([self.b])),
        ]
            .concat()
    }
}
impl From<SomeRelationWithPublicInput> for SomeRelationWithoutInput {
    fn from(obj: SomeRelationWithPublicInput) -> Self {
        Self { a: obj.a }
    }
}
pub struct SomeRelationWithFullInput {
    a: u8,
    b: CF,
    d: u8,
    c: u64,
}
impl SomeRelationWithFullInput {
    pub fn new(a: u8, b: u16, d: u8, c: u32) -> Self {
        Self {
            a,
            b: parse_u16(b),
            d,
            c: c.into(),
        }
    }
    pub fn a(&self) -> &u8 {
        &self.a
    }
    pub fn b(
        &self,
    ) -> core::result::Result<&CF, ark_relations::r1cs::SynthesisError> {
        Ok(&self.b)
    }
    pub fn d(
        &self,
    ) -> core::result::Result<&u8, ark_relations::r1cs::SynthesisError> {
        Ok(&self.d)
    }
    pub fn c(
        &self,
    ) -> core::result::Result<&u64, ark_relations::r1cs::SynthesisError> {
        Ok(&self.c)
    }
}
impl From<SomeRelationWithFullInput> for SomeRelationWithPublicInput {
    fn from(obj: SomeRelationWithFullInput) -> Self {
        Self {
            a: obj.a,
            b: obj.b,
            d: obj.d,
        }
    }
}
impl<Field: ark_ff::PrimeField> ark_relations::r1cs::ConstraintSynthesizer<Field>
for SomeRelationWithoutInput {
    fn generate_constraints(
        self,
        cs: ark_relations::r1cs::ConstraintSystemRef<Field>,
    ) -> ark_relations::r1cs::Result<()> {
        if cs.is_in_setup_mode() {
            Ok(())
        } else {
            #[cfg(feature = "std")]
            {
                {
                    ::std::io::_eprint(
                        ::core::fmt::Arguments::new_v1(
                            &[
                                "For proof generation, you should use relation object with full input.\n",
                            ],
                            &[],
                        ),
                    );
                };
            }
            Err(ark_relations::r1cs::SynthesisError::AssignmentMissing)
        }
    }
}
impl<Field: ark_ff::PrimeField> ark_relations::r1cs::ConstraintSynthesizer<Field>
for SomeRelationWithFullInput {
    fn generate_constraints(
        self,
        cs: ark_relations::r1cs::ConstraintSystemRef<Field>,
    ) -> ark_relations::r1cs::Result<()> {
        Ok(())
    }
}
```

### CODE REVIEW

File: bin/cliain/src/snark_relations/relations.rs [hunk 1]
```diff
@@ -273,8 +275,8 @@ impl GetPublicInput<CircuitField> for RelationArgs {
                 ..
             } => match (note, token_id, token_amount) {
                 (Some(note), Some(token_id), Some(token_amount)) => {
-                    DepositRelation::with_public_input(*note, *token_id, *token_amount)
-                        .public_input()
+                    DepositRelationWithPublicInput::new(*note, *token_id, *token_amount)
+                        .serialize_public_input()
```
Comment (deuszx) [inline bin/cliain/src/snark_relations/relations.rs:279]: Could this be called simply _serialize_? We already have the `*WithPublicInput` in the type name.

File: relations/src/shielder/deposit.rs [hunk 2]
```diff
@@ -1,149 +1,96 @@
...
-    },
-};
+use ark_relations::ns;
+use snark_relation_proc_macro::snark_relation;
+
+use crate::{BackendNote, FrontendNote};
+
+fn convert_note(front: FrontendNote) -> BackendNote {
+    BackendNote::from(BigInteger256::new(front))
+}
```
Comment (deuszx) [inline relations/src/shielder/deposit.rs:10]: Could this be turned into an `Into`/`From` implementation?

And, by extend, would it work if we had `mod backend { ... }` and `mod frontend { ... }` with representations and mappings (as `From`/`Into` impls) between the two? Then, maybe we could cut down on the "boilerplate" in the `#[public_input(...)]` macro and just require that there exists `BackendType: Into<FrontedType>` instead of `#[public_input(frontend_type = "FrontendNote", parse_with = "convert_note")]`

File: bin/cliain/src/snark_relations/relations.rs (hunk 1 above)
Comment (pmikolajczyk41) [reply bin/cliain/src/snark_relations/relations.rs:279]: I'm slightly against - `serialize` on its own would indicate that we are serializing whole `RelationWithPublicInput` object, which might include constants and other irrelevant things

Comment (deuszx) [reply bin/cliain/src/snark_relations/relations.rs:279]: OK, I'm convinced :)

File: relations/src/shielder/deposit.rs (hunk 2 above)
Comment (pmikolajczyk41) [reply relations/src/shielder/deposit.rs:10]: into/from won't work, since we cannot implement foreign traits for foreign types (notice that all front/backend types are just aliases)

I was trying to make a step towards it here: https://github.com/Cardinal-Cryptography/HouseSnark/pull/35/files, but back then there was _a lot_ of code to implement strong typing in full... however, I would revisit this somewhere soon

Comment (deuszx) [reply relations/src/shielder/deposit.rs:10]: We could add our own mapping trait then and implement that instead (of `Into`/`From`).

Comment (pmikolajczyk41) [reply relations/src/shielder/deposit.rs:10]: yeap, that's the easiest alternative, however, slightly out of scope here

Comment (deuszx) [reply relations/src/shielder/deposit.rs:10]: Yes, just mentioning the possible improvement to the macro.

Comment (pmikolajczyk41) [reply relations/src/shielder/deposit.rs:10]: talking about improvements, it should be fairly easy to add generics support as well

File: relations/src/xor.rs [hunk 3]
```diff
@@ -21,69 +7,86 @@ use crate::{
...
-        Self {
-            public_xoree: Some(public_xoree),
-            private_xoree: None,
-            result,
-            _phantom: PhantomData,
-        }
-    }
-}
+    #[circuit_field]
+    pub type CircuitField = crate::CircuitField;
```
Comment (deuszx) [inline relations/src/xor.rs:17]: Do you expect we would ever have a different type for this associated type?

Comment (pmikolajczyk41) [reply relations/src/xor.rs:17]: :man_shrugging: I really don't know...

**Review by mike1729 — APPROVED** [review]
Treat this as soft approval as I've never written such macro.

File: relations/src/shielder/deposit.rs [hunk 4]
```diff
@@ -1,149 +1,96 @@
...
-        })?;
-
-        let trapdoor = FpVar::new_witness(ns!(cs, "trapdoor"), || {
-            self.trapdoor.ok_or(AssignmentMissing)
-        })?;
-        let nullifier = FpVar::new_witness(ns!(cs, "nullifier"), || {
-            self.nullifier.ok_or(AssignmentMissing)
-        })?;
+        let trapdoor = FpVar::new_witness(ns!(cs, "trapdoor"), || self.trapdoor())?;
+        let nullifier = FpVar::new_witness(ns!(cs, "nullifier"), || self.nullifier())?;
```
Comment (deuszx) [inline relations/src/shielder/deposit.rs:55]: I can see yet another chance for improvement here: we know which fields are public (and private) thanks to the macro - we can generate the "setup" (all the `FpVar::new_input` and `FpVar::new_witness` calls) in the macro itself and "just" require the developer to implement the validity checks.

Comment (pmikolajczyk41) [reply relations/src/shielder/deposit.rs:55]: this would require some consideration how to handle it comfortably - sometimes, you have e.g. a vector of `FpVar`; also, the order of declaration is important

anyway - yes, you're right and we can think about it soon

**Review by maciejzelaszczyk — APPROVED** [review]
This is a behemoth amount of work. Technically, very impressive. I find that it considerably simplifies the creation of relations. On the flip side, we need to be cognizant of the obfuscation the macros introduce. It might bite us in the not so distant future as there certainly is a complexity cost. Having said that, let us try the macro route and see how far it takes us.

TLDR: LGTM.

**Review by deuszx — APPROVED** [review]
Titanic work, amazing job. People will be writing poems about it 👏 🙇 

I'm not gonna lie, _I don't understand half of it half as well as I should understand; and I appreciate less than half of it half as well as it deserves._

Or to put more simply: I don't understand everything and I can't yet see how it all will be used. I've  left some comments though.

File: relations/src/proc_macro/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+///     where `<R>` is the name of the relation object struct. The first one will have only
+///     constants as its fields, the second one will have additionally public inputs, and the last
+///     one will have all the data.
+///  -  `new(..)` constructors for every struct. **Important**: the order of constructor arguments
+///     is: all the constants, then public inputs, and at the end private inputs. The order in each
+///     group is inherited from the relation object definition.
+///  -  Getters for the fields. For constants, the signature is `fn <field>(&self) ->
+///     &<field_type>`. For public and private inputs, the signature is `fn <field>(&self) ->
+///     Result<&<field_type>, SynthesisError>`. All the structs have the same set of getters.
+///     When a field is missing, `SynthesisError::MissingAssignment` is returned.
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/lib.rs:40]: all the structs have the same set of getters, but different sets of fields - e.g. WithoutInput won't have any of the public or private input - in that case we return error; this logic is needed for uniform circuit definition

File: relations/src/proc_macro/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+///  -  Conversions from `<R>WithFullInput` to `<R>WithPublicInput` and from `<R>WithPublicInput`
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/lib.rs:41]: I really don't see a place for that - there's no API that requires `WithoutInput` and won't accept `WithFullInput` (and work differently)

with `WithPublicInput` the situation is different - if you have `WithFullInput`, you are not able to serialize your public inputs; and if you have `WithPublicInput`, you cannot generate keys

File: relations/src/proc_macro/src/parse_utils.rs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                        let path = path.get_ident().map(|i| i.to_string()).ok_or(err.clone())?;
+                        if !permissible_config.contains(path.as_str()) {
+                            return Err(err);
+                        }
+                        let lit = match lit {
+                            Lit::Str(lit_str) => lit_str.value(),
+                            Lit::Int(lit_int) => lit_int.token().to_string(),
+                            _ => return Err(err),
+                        };
+                        config.insert(path, lit);
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/parse_utils.rs:83]: unit testing macro functions is not that nice :| this function just parses `attr1 = "val1", attr2 = "val2"` into a hashmap, with that restriction that all `attr1`, `attr2` must belong to `permissible_config`

File: relations/src/proc_macro/src/code_generation.rs [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+/// Generates struct, constructor and getters for the relation object with constants only.
+fn generate_relation_without_input(ir: &IR) -> SynResult<TokenStream2> {
+    let struct_name = struct_name_without_input(&ir.relation_base_name);
+    let const_frontend_decls = field_frontend_decls(&ir.constants);
+    let const_backend_decls = field_backend_decls(&ir.constants);
+    let const_castings = field_castings(&ir.constants)?;
+    let getters = [
+        plain_field_getters(&ir.constants),
+        failing_field_getters(&ir.public_inputs),
+        failing_field_getters(&ir.private_inputs),
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/code_generation.rs:42]: yes I need - when we call `<R>WithoutInput`.generate_constraints(cs)`, then these getters are used in the methods like `new_witness`, `new_input`; although they won't be actually called, we must provide properly typed closure; and this way we are consistent across all relation object variations

File: relations/src/proc_macro/src/code_generation.rs [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    ]
+    .concat();
+    let castings = [
+        field_castings(&ir.constants)?,
+        field_castings(&ir.public_inputs)?,
+    ]
+    .concat();
+    let getters = [
+        plain_field_getters(&ir.constants),
+        successful_field_getters(&ir.public_inputs),
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/code_generation.rs:137]: the same reason as above

File: relations/src/proc_macro/src/code_generation.rs [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+        failing_field_getters(&ir.private_inputs),
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/code_generation.rs:138]: as above

File: relations/src/proc_macro/src/code_generation.rs [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+            pub fn new(#(#const_frontend_decls),*) -> Self {
+                Self { #(#const_castings),* }
+            }
+            #(#getters)*
+        }
+    })
+}
+
+/// Returns reordered copies of the elements in `fields`.
+fn get_ordered_inputs(fields: &[PublicInputField]) -> SynResult<Vec<PublicInputField>> {
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/code_generation.rs:60]: removed anyway

File: relations/src/proc_macro/src/intermediate_representation.rs [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+/// Tries casting every element in `fields` into `F`.
+fn cast_fields<F: TryFrom<Field, Error = SynError>>(fields: Vec<Field>) -> SynResult<Vec<F>> {
+    fields
+        .into_iter()
+        .map(TryInto::<F>::try_into)
+        .collect::<Vec<_>>()
+        .into_iter()
+        .collect()
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/intermediate_representation.rs]: huh, I was sure that it didn't work for me for the first time... nvm, changed

File: relations/src/proc_macro/src/intermediate_representation.rs [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+                "Invalid module - it is expected to be inlined",
+            )
+        })?
+        .1;
+
+    let span = item_mod.span();
+
+    let struct_def = extract_item(items, as_relation_object_def, span, "relation object")?;
+    let circuit_def = extract_item(items, as_circuit_def, span, "circuit definition")?;
+    let circuit_field = extract_item(items, as_circuit_field_def, span, "circuit field")?;
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/intermediate_representation.rs]: done

File: relations/src/proc_macro/src/intermediate_representation.rs [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+
+    Ok(Items {
+        struct_def,
+        circuit_def,
+        circuit_field,
+        imports,
+    })
+}
+
+/// Returns all the elements of `fields` that are attributed with `field_name`, e.g.
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/intermediate_representation.rs]: done

File: relations/src/proc_macro/src/intermediate_representation.rs [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+        circuit_field
+            .attrs
+            .retain(|a| !a.path.is_ident(CIRCUIT_FIELD_DEF));
+
+        // Extract all fields. There should be at least one field. All fields must be named.
+        let fields = match struct_def.fields {
+            Fields::Named(fields) => Ok(fields),
+            _ => Err(SynError::new(
+                struct_def.fields.span(),
+                "The relation should have some fields and they should be named.",
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/intermediate_representation.rs]: done

File: relations/src/proc_macro/src/intermediate_representation.rs [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+            if !matches!(struct_def.vis, Visibility::Inherited) {
+                eprintln!(
+                    "Warning: The `{relation_base_name}` struct is public, but will be erased."
+                )
+            };
+            if !matches!(circuit_definition.vis, Visibility::Inherited) {
+                eprintln!("Warning: The circuit definition is public, but will be erased.")
+            }
+            if !matches!(circuit_field.vis, Visibility::Public(_)) {
+                eprintln!("Warning: The circuit field must be public. Visibility will be changed.");
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/intermediate_representation.rs]: removed type definition as discussed offline

File: relations/src/proc_macro/src/intermediate_representation.rs [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+                parse_with: config.get(FIELD_PARSER).cloned(),
+            },
+            serialize_with: config.get(FIELD_SERIALIZER).cloned(),
+            order,
+        })
+    }
+}
+
+/// Intermediate representation of the source code.
+pub(super) struct IR {
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/intermediate_representation.rs]: moved

File: relations/src/proc_macro/src/generation_utils.rs [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+use syn::{spanned::Spanned, Error as SynError, Result as SynResult};
+
+use crate::intermediate_representation::{PublicInputField, RelationField};
+
+/// Forcily extracts ident from the field.
+fn get_ident(f: &RelationField) -> &Ident {
+    f.field
+        .ident
+        .as_ref()
+        .expect("We are working on named fields")
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/generation_utils.rs]: done

File: relations/src/proc_macro/src/generation_utils.rs [hunk 19]
```diff
@@ -0,0 +1,167 @@
+use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
+use quote::quote;
+use syn::{spanned::Spanned, Error as SynError, Result as SynResult};
+
+use crate::intermediate_representation::{PublicInputField, RelationField};
+
+/// Forcily extracts ident from the field.
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/generation_utils.rs]: done

File: relations/src/proc_macro/src/intermediate_representation.rs [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+
+/// Full data for public inputs.
+#[derive(Clone)]
+pub(super) struct PublicInputField {
+    /// Common data for all inputs.
+    pub inner: RelationField,
+    /// The value of the `serialize_with` modifier, if any.
+    pub serialize_with: Option<String>,
+    /// The value of the `order` modifier, if any.
+    pub order: Option<usize>,
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/intermediate_representation.rs]: removed as discussed offline

File: relations/src/proc_macro/src/parse_utils.rs [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+        0 => Err(SynError::new(
+            field.span(),
+            "Relation field should be one of 3 types (constant, public, private).",
+        )),
+        1 => Ok(attrs[0]),
+        _ => Err(SynError::new(
+            attrs[1].span(),
+            "Relation field cannot have two types.",
+        )),
+    }
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/proc_macro/src/parse_utils.rs]: sure, done

File: relations/src/shielder/deposit.rs [hunk 22]
```diff
@@ -1,149 +1,96 @@
...
+    #[relation_object_definition]
+    struct DepositRelation {
+        #[public_input(frontend_type = "FrontendNote", parse_with = "convert_note")]
+        pub note: BackendNote,
+        #[public_input(frontend_type = "FrontendTokenId")]
+        pub token_id: BackendTokenId,
+        #[public_input(frontend_type = "FrontendTokenAmount")]
+        pub token_amount: BackendTokenAmount,
+
+        // Private inputs
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply relations/src/shielder/deposit.rs]: dropped

**Review by deuszx — APPROVED** [review]
